File: src/security/SecurityHeaders.ts
// Security Headers
// Defines and validates security headers, CORS policy, and CSP rules

export interface SecurityHeaderConfig {
  contentSecurityPolicy: Record<string, string[]>;
  corsAllowedOrigins: string[];
  corsAllowedMethods: string[];
  hstsMaxAge: number;
  frameOptions: 'DENY' | 'SAMEORIGIN';
  referrerPolicy: string;
  permissionsPolicy: Record<string, string[]>;
}

/**
 * Security Headers Manager
 * Generates recommended HTTP security headers and validates configuration
 */
export class SecurityHeadersManager {
  private config: SecurityHeaderConfig;

  constructor(customConfig?: Partial<SecurityHeaderConfig>) {
    this.config = {
      contentSecurityPolicy: {
        'default-src': ["'self'"],
        'script-src': ["'self'", "'strict-dynamic'"],
        'style-src': ["'self'", "'unsafe-inline'"],
        'img-src': ["'self'", 'data:', 'https:'],
        'connect-src': ["'self'"],
        'font-src': ["'self'"],
        'frame-ancestors': ["'none'"],
        'base-uri': ["'self'"],
        'form-action': ["'self'"],
      },
      corsAllowedOrigins: ['https://kotravel.com', 'https://app.kotravel.com'],
      corsAllowedMethods: ['GET', 'POST', 'PUT', 'DELETE', 'OPTIONS'],
      hstsMaxAge: 31536000, // 1 year
      frameOptions: 'DENY',
      referrerPolicy: 'strict-origin-when-cross-origin',
      permissionsPolicy: {
        camera: ['self'],
        microphone: [],
        geolocation: ['self'],
        payment: [],
      },
      ...customConfig,
    };
  }

  /**
   * Generate all recommended security headers
   */
  generateHeaders(): Record<string, string> {
    return {
      'Content-Security-Policy': this.buildCSPHeader(),
      'Strict-Transport-Security': `max-age=${this.config.hstsMaxAge}; includeSubDomains; preload`,
      'X-Frame-Options': this.config.frameOptions,
      'X-Content-Type-Options': 'nosniff',
      'Referrer-Policy': this.config.referrerPolicy,
      'Permissions-Policy': this.buildPermissionsPolicyHeader(),
      'X-XSS-Protection': '1; mode=block',
      'X-DNS-Prefetch-Control': 'off',
    };
  }

  /**
   * Validate if origin is allowed for CORS
   */
  isOriginAllowed(origin: string): boolean {
    return this.config.corsAllowedOrigins.includes(origin);
  }

  /**
   * Generate CORS headers for a validated origin
   */
  generateCORSHeaders(origin: string): Record<string, string> | null {
    if (!this.isOriginAllowed(origin)) return null;

    return {
      'Access-Control-Allow-Origin': origin,
      'Access-Control-Allow-Methods': this.config.corsAllowedMethods.join(', '),
      'Access-Control-Allow-Headers': 'Content-Type, Authorization, X-Webhook-Signature',
      'Access-Control-Allow-Credentials': 'true',
      'Access-Control-Max-Age': '86400',
    };
  }

  /**
   * Add allowed origin
   */
  addAllowedOrigin(origin: string): void {
    if (!this.config.corsAllowedOrigins.includes(origin)) {
      this.config.corsAllowedOrigins.push(origin);
    }
  }

  /**
   * Remove allowed origin
   */
  removeAllowedOrigin(origin: string): void {
    this.config.corsAllowedOrigins = this.config.corsAllowedOrigins.filter(o => o !== origin);
  }

  /**
   * Add CSP directive source
   */
  addCSPSource(directive: string, source: string): void {
    if (!this.config.contentSecurityPolicy[directive]) {
      this.config.contentSecurityPolicy[directive] = [];
    }
    if (!this.config.contentSecurityPolicy[directive].includes(source)) {
      this.config.contentSecurityPolicy[directive].push(source);
    }
  }

  /**
   * Validate a request against security policy
   */
  validateRequest(headers: Record<string, string>): { valid: boolean; issues: string[] } {
    const issues: string[] = [];

    const origin = headers['origin'] || headers['Origin'];
    if (origin && !this.isOriginAllowed(origin)) {
      issues.push(`Origin '${origin}' is not in allowlist`);
    }

    const contentType = headers['content-type'] || headers['Content-Type'];
    if (contentType && !contentType.includes('application/json') && !contentType.includes('multipart/form-data')) {
      issues.push(`Unexpected content-type: ${contentType}`);
    }

    return { valid: issues.length === 0, issues };
  }

  /**
   * Get current configuration
   */
  getConfig(): SecurityHeaderConfig {
    return { ...this.config };
  }

  // ==================== PRIVATE METHODS ====================

  private buildCSPHeader(): string {
    return Object.entries(this.config.contentSecurityPolicy)
      .map(([directive, sources]) => `${directive} ${sources.join(' ')}`)
      .join('; ');
  }

  private buildPermissionsPolicyHeader(): string {
    return Object.entries(this.config.permissionsPolicy)
      .map(([feature, allowlist]) => {
        const sources = allowlist.map(s => (s === 'self' ? 'self' : `"${s}"`)).join(' ');
        return `${feature}=(${sources})`;
      })
      .join(', ');
  }
}

let securityHeaders: SecurityHeadersManager | null = null;

export function getSecurityHeaders(): SecurityHeadersManager {
  if (!securityHeaders) {
    securityHeaders = new SecurityHeadersManager();
  }
  return securityHeaders;
}

File: backend/security_headers.py
"""
Security headers middleware and utilities.
Implements OWASP recommended security headers.
"""

from flask import Flask, request, jsonify
from functools import wraps
from typing import Dict, Any

def add_security_headers(app: Flask) -> None:
    """
    Add security headers to all responses.

    Args:
        app: Flask application instance
    """

    @app.after_request
    def set_security_headers(response):
        """Add security headers to response."""

        # Prevent MIME type sniffing
        response.headers['X-Content-Type-Options'] = 'nosniff'

        # Enable XSS protection (modern browsers ignore, but good for legacy support)
        response.headers['X-XSS-Protection'] = '1; mode=block'

        # Prevent clickjacking attacks (embedding in iframe)
        response.headers['X-Frame-Options'] = 'DENY'

        # Control how much referrer information is sent
        response.headers['Referrer-Policy'] = 'strict-origin-when-cross-origin'

        # Content Security Policy (CSP)
        # Restrict where content can be loaded from
        csp_header = (
            "default-src 'self'; "
            "script-src 'self' 'unsafe-inline'; "  # unsafe-inline for development only
            "style-src 'self' 'unsafe-inline'; "
            "img-src 'self' data: https:; "
            "font-src 'self'; "
            "connect-src 'self'; "
            "frame-ancestors 'none'; "
            "base-uri 'self'; "
            "form-action 'self';"
        )
        response.headers['Content-Security-Policy'] = csp_header

        # HSTS (HTTP Strict Transport Security)
        # Force HTTPS in production
        if request.environ.get('wsgi.url_scheme') == 'https':
            response.headers['Strict-Transport-Security'] = (
                'max-age=31536000; includeSubDomains; preload'
            )

        # Permissions-Policy (formerly Feature-Policy)
        # Control which browser features can be used
        response.headers['Permissions-Policy'] = (
            'geolocation=(), '
            'microphone=(), '
            'camera=(), '
            'payment=(), '
            'usb=(), '
            'magnetometer=(), '
            'gyroscope=(), '
            'accelerometer=()'
        )

        # Additional security headers
        response.headers['X-Permitted-Cross-Domain-Policies'] = 'none'
        response.headers['Cross-Origin-Resource-Policy'] = 'same-origin'
        response.headers['Cross-Origin-Opener-Policy'] = 'same-origin'
        response.headers['Cross-Origin-Embedder-Policy'] = 'require-corp'

        # Remove unnecessary headers
        response.headers.pop('Server', None)
        response.headers.pop('X-Powered-By', None)

        return response


def require_secure_headers(f):
    """
    Decorator to require specific security headers in request.
    Used for sensitive operations (state-changing requests).
    """
    @wraps(f)
    def decorated_function(*args, **kwargs):
        # Check Content-Type for POST/PUT requests
        if request.method in ['POST', 'PUT', 'DELETE', 'PATCH']:
            content_type = request.headers.get('Content-Type', '')
            if not content_type.startswith('application/json'):
                return jsonify({'error': 'Content-Type must be application/json'}), 400

        return f(*args, **kwargs)

    return decorated_function


def validate_request_origin(f):
    """
    Decorator to validate request origin.
    Ensures requests come from allowed origins.
    """
    @wraps(f)
    def decorated_function(*args, **kwargs):
        origin = request.headers.get('Origin')
        referer = request.headers.get('Referer')

        # Allow requests without Origin/Referer from same-origin
        if request.method in ['POST', 'PUT', 'DELETE', 'PATCH']:
            if not origin and not referer:
                # Same-origin requests may not have Origin header
                pass
            elif origin:
                # Validate origin in CORS middleware
                pass

        return f(*args, **kwargs)

    return decorated_function


class SecurityHeaderConfig:
    """Static security header configurations."""

    # CSP directives for different content types
    CSP_DIRECTIVES = {
        'default-src': ["'self'"],
        'script-src': ["'self'", "'unsafe-inline'"],  # Inline for development only
        'style-src': ["'self'", "'unsafe-inline'"],
        'img-src': ["'self'", "data:", "https:"],
        'font-src': ["'self'"],
        'connect-src': ["'self'"],
        'frame-ancestors': ["'none'"],
        'base-uri': ["'self'"],
        'form-action': ["'self'"],
        'upgrade-insecure-requests': [],
    }

    @staticmethod
    def get_csp_header(custom_directives: Dict[str, list] = None) -> str:
        """
        Generate CSP header string.

        Args:
            custom_directives: Override default directives

        Returns:
            CSP header string
        """
        directives = SecurityHeaderConfig.CSP_DIRECTIVES.copy()

        if custom_directives:
            directives.update(custom_directives)

        parts = []
        for key, values in directives.items():
            if values:
                parts.append(f"{key} {' '.join(values)}")
            else:
                parts.append(key)

        return "; ".join(parts)

    @staticmethod
    def get_hsts_header(
        max_age: int = 31536000,
        include_subdomains: bool = True,
        preload: bool = True
    ) -> str:
        """
        Generate HSTS header string.

        Args:
            max_age: Max age in seconds (default 1 year)
            include_subdomains: Include subdomains
            preload: Enable HSTS preload

        Returns:
            HSTS header string
        """
        parts = [f"max-age={max_age}"]

        if include_subdomains:
            parts.append("includeSubDomains")

        if preload:
            parts.append("preload")

        return "; ".join(parts)


def remove_sensitive_headers(response) -> None:
    """
    Remove headers that expose server information.

    Args:
        response: Flask response object
    """
    sensitive_headers = [
        'Server',
        'X-Powered-By',
        'X-AspNet-Version',
        'X-Runtime-Version',
        'X-Content-Type-Options',  # Keep for security
    ]

    for header in sensitive_headers:
        response.headers.pop(header, None)
